### PR TITLE
stop sessions

### PR DESCRIPTION
**Pull request by itayd**
- implement session stop, graceful and non-graceful.
- add STOPPED state

### CODE REVIEW

File: internal/backend/sessions/sessionworkflows/workflow.go [hunk 1]
```diff
@@ -99,6 +100,17 @@ func (w *sessionWorkflow) updateState(ctx workflow.Context, state sdktypes.Sessi
 
 	w.state = state
 
+	if ctx.Err() == workflow.ErrCanceled {
+		goCtx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
```
Comment (ig-ra) [inline internal/backend/sessions/sessionworkflows/workflow.go]: let's define this as a const somewhere

File: internal/backend/sessionsgrpcsvc/svc.go [hunk 2]
```diff
@@ -71,6 +71,25 @@ func (s *server) Get(ctx context.Context, req *connect.Request[sessionsv1.GetReq
 	return connect.NewResponse(&sessionsv1.GetResponse{Session: session.ToProto()}), nil
 }
 
+func (s *server) Stop(ctx context.Context, req *connect.Request[sessionsv1.StopRequest]) (*connect.Response[sessionsv1.StopResponse], error) {
```
Comment (ig-ra) [inline internal/backend/sessionsgrpcsvc/svc.go:74]: how stopped sesstions are handled in DB?

**Review by ig-ra — COMMENTED** [review]
I think CANCELLED state is more like FINISHED state, not ERROR. But maybe we should keep 3 different states for simplicity and clarity

File: internal/backend/sessionsgrpcsvc/svc.go (hunk 2 above)
Comment (itayd) [reply internal/backend/sessionsgrpcsvc/svc.go:74]: not sure i follow your question here

**Comment by itayd** [comment]
added STOPPED state

File: internal/backend/sessions/sessionworkflows/workflow.go (hunk 1 above)
Comment (itayd) [reply internal/backend/sessions/sessionworkflows/workflow.go]: done

File: internal/backend/sessionsgrpcsvc/svc.go (hunk 2 above)
Comment (ig-ra) [reply internal/backend/sessionsgrpcsvc/svc.go:74]: I meant that there are db functions that should be aware of this change, e.g. listDeploymentsWithStats.

File: internal/backend/sessions/sessionworkflows/workflow.go [hunk 3]
```diff
@@ -28,6 +29,8 @@ import (
 const (
 	envVarsModuleName     = "env"
 	integrationPathPrefix = "@"
+
+	limittedTimeout = 5 * time.Second
```
Comment (ig-ra) [inline internal/backend/sessions/sessionworkflows/workflow.go]: let's open a ticket to extract all time consts in some file?
There are more - in sessions, in tests etc...

File: sdk/sdktypes/session_state.go [hunk 4]
```diff
@@ -18,6 +18,7 @@ type SessionStateTraits struct{}
 
 func (SessionStateTraits) Validate(m *SessionStatePB) error {
 	return errors.Join(
+		objectField[SessionStateStopped]("canceled", m.Stopped),
```
Comment (ig-ra) [inline sdk/sdktypes/session_state.go]: naming inconsistency? is session stopped or cancelled? let's use the same naming?

File: internal/backend/sessions/sessionworkflows/workflow.go [hunk 5]
```diff
@@ -480,7 +479,9 @@ func (w *sessionWorkflow) run(wctx workflow.Context) (prints []string, err error
 
 	state := sdktypes.NewSessionStateCompleted(prints, run.Values(), retVal)
 
-	w.updateState(wctx, state)
+	if err := w.updateState(wctx, state); err != nil {
```
Comment (ig-ra) [inline internal/backend/sessions/sessionworkflows/workflow.go]:
```suggestion
	return prints, w.updateState(wctx, state)
```

File: internal/backend/db/dbgorm/sessions.go [hunk 6]
```diff
@@ -110,7 +110,14 @@ func addSessionLogRecord(tx *gorm.DB, sessionID string, logr sdktypes.SessionLog
 
 func (db *gormdb) AddSessionPrint(ctx context.Context, sessionID sdktypes.SessionID, print string) error {
 	return translateError(
-		addSessionLogRecord(db.db, sessionID.String(), sdktypes.NewPrintSessionLogRecord(print)))
+		addSessionLogRecord(db.db, sessionID.String(), sdktypes.NewPrintSessionLogRecord(print)),
```
Comment (ig-ra) [inline internal/backend/db/dbgorm/sessions.go:113]: Let's add tests? at least to check that all aspects of stopped session is working as expected?
e.g. in listDeploymentsWIthStats and other places?

**Review by ig-ra — APPROVED** [review]

**Comment by ig-ra** [comment]
preliminary approval. @itayd  please review remaining comments

File: internal/backend/sessionsgrpcsvc/svc.go (hunk 2 above)
Comment (itayd) [reply internal/backend/sessionsgrpcsvc/svc.go:74]: ah, i think that is the only one? fixed.

File: sdk/sdktypes/session_state.go (hunk 4 above)
Comment (itayd) [reply sdk/sdktypes/session_state.go]: fixed

File: internal/backend/sessions/sessionworkflows/workflow.go (hunk 5 above)
Comment (itayd) [reply internal/backend/sessions/sessionworkflows/workflow.go]: done

File: internal/backend/sessions/sessionworkflows/workflow.go (hunk 3 above)
Comment (itayd) [reply internal/backend/sessions/sessionworkflows/workflow.go]: i unified all of them

File: internal/backend/db/db.go [hunk 7]
```diff
@@ -135,6 +135,7 @@ type DB interface {
 	GetSessionLog(ctx context.Context, sessionID sdktypes.SessionID) (sdktypes.SessionLog, error)
 	UpdateSessionState(ctx context.Context, sessionID sdktypes.SessionID, state sdktypes.SessionState) error
 	AddSessionPrint(ctx context.Context, sessionID sdktypes.SessionID, print string) error
+	AddSessionStopRequested(ctx context.Context, sessionID sdktypes.SessionID, reason string) error
```
Comment (daabr) [inline internal/backend/db/db.go]: Just my 2 cents - "add session stop requested" is a very weird function name, for multiple reasons:

- Multiple verbs, without an obvious meaning
- "Requested" implies an event/state, but this is a function

File: internal/backend/sessions/sessionworkflows/workflow.go [hunk 8]
```diff
@@ -28,8 +29,14 @@ import (
 const (
 	envVarsModuleName     = "env"
 	integrationPathPrefix = "@"
+
+	limittedTimeout = 5 * time.Second
```
Comment (daabr) [inline internal/backend/sessions/sessionworkflows/workflow.go]:
```suggestion
	limitedTimeout = 5 * time.Second
```

File: internal/backend/sessions/sessionworkflows/workflow.go [hunk 9]
```diff
@@ -28,8 +29,14 @@ import (
...
 )
 
+func withLimittedTimeout(ctx context.Context) (context.Context, func()) {
+	return context.WithTimeout(ctx, limittedTimeout)
```
Comment (daabr) [inline internal/backend/sessions/sessionworkflows/workflow.go]:
```suggestion
func withLimitedTimeout(ctx context.Context) (context.Context, func()) {
	return context.WithTimeout(ctx, limitedTimeout)
```

File: sdk/sdktypes/session_state_type.go [hunk 10]
```diff
@@ -38,7 +39,9 @@ func SessionStateTypeFromProto(e sessionsv1.SessionStateType) (SessionStateType,
 func ParseSessionStateType(raw string) (SessionStateType, error) {
 	return ParseEnum[SessionStateType](raw)
 }
+
 func (e SessionStateType) IsFinal() bool {
 	return e.v == sessionsv1.SessionStateType_SESSION_STATE_TYPE_ERROR ||
-		e.v == sessionsv1.SessionStateType_SESSION_STATE_TYPE_COMPLETED
+		e.v == sessionsv1.SessionStateType_SESSION_STATE_TYPE_COMPLETED ||
```
Comment (daabr) [inline sdk/sdktypes/session_state_type.go:45]: Please also update these:

- https://github.com/autokitteh/autokitteh/blob/a3cf7994093e3ffb79acbd95f48a6e2332028b57/tests/system/README.md?plain=1#L134-L139
- https://github.com/autokitteh/autokitteh/blob/a3cf7994093e3ffb79acbd95f48a6e2332028b57/tests/system/client.go#L76-L77

File: internal/backend/sessions/sessionworkflows/workflows.go [hunk 11]
```diff
@@ -119,64 +120,111 @@ func (ws *workflows) getSessionData(ctx workflow.Context, sessionID sdktypes.Ses
...
 
 	if depID := data.Session.DeploymentID(); depID.IsValid() {
-		workflow.Go(ctx, func(ctx workflow.Context) {
-			// TODO: can this be done async? do we care?
-			if err := workflow.ExecuteLocalActivity(
-				ctx,
-				ws.deactivateDrainedDeployment,
-				depID,
-			).Get(ctx, nil); err != nil {
+		// We cannot rely on workflow context here as it might have been cancelled.
```
Comment (daabr) [inline internal/backend/sessions/sessionworkflows/workflows.go]: Nit: I see you already fixed the code (cancelled --> canceled), please fix this comment too, and also in line 142 below.

File: internal/backend/db/dbgorm/sessions.go (hunk 6 above)
Comment (itayd) [reply internal/backend/db/dbgorm/sessions.go:113]: too complicated for now - will do in a future PR

File: internal/backend/db/db.go (hunk 7 above)
Comment (itayd) [reply internal/backend/db/db.go]: tbh it does its job. if you have a better name lmk :) I will change it to `AddSessionStopRequest` though.

File: internal/backend/sessions/sessionworkflows/workflow.go (hunk 9 above)
Comment (itayd) [reply internal/backend/sessions/sessionworkflows/workflow.go]: fixed

File: internal/backend/sessions/sessionworkflows/workflows.go (hunk 11 above)
Comment (itayd) [reply internal/backend/sessions/sessionworkflows/workflows.go]: done

File: sdk/sdktypes/session_state_type.go (hunk 10 above)
Comment (itayd) [reply sdk/sdktypes/session_state_type.go:45]: done

**Pull request merged**